Test suite fails with non-UTF8 locale (LC_ALL=C)

The Debian reproducible-builders vary environment variables to see if packages are reproducible with variations in things like `HOME`, `LANG`, `LC_ALL`, etc. Reproducible builds are part of the software supply chain security story - lots more on [https://reproducible-builds.org/](https://reproducible-builds.org/)

A few tests in the test suite for translate-toolkit fail with `LC_ALL=C` but are fine with `LC_ALL=C.UTF8`:
```
FAILED tests/translate/storage/test_po.py::TestPOFile::test_wrapping_cjk
FAILED tests/translate/storage/test_po.py::TestPOFile::test_wrap_parenthesis
FAILED tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrapping_cjk
FAILED tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrap_parenthesis
```

A simple reproducer:
```
$ LC_ALL=C python3.13 -m pytest -v tests/translate/storage/test_po.py::TestPOFile::test_wrapping_cjk tests/translate/storage/test_po.py::TestPOFile::test_wrap_parenthesis tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrapping_cjk tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrap_parenthesis
```

(fails in both Debian bookworm, trixie and sid)

--- 
Full output
```console
$ LC_ALL=C python3 -m pytest -v tests/translate/storage/test_po.py::TestPOFile::test_wrapping_cjk tests/translate/storage/test_po.py::TestPOFile::test_wrap_parenthesis tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrapping_cjk tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrap_parenthesis
============================================== test session starts ==============================================
platform linux -- Python 3.13.5, pytest-8.3.5, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: …/translate-toolkit/upstream
configfile: pyproject.toml
plugins: limnoria_plugin-0.1.0, mock-3.14.0, cov-5.0.0, Faker-33.3.1, anyio-4.8.0, syrupy-4.9.1, typeguard-4.4.2
collected 4 items                                                                                               

tests/translate/storage/test_po.py::TestPOFile::test_wrapping_cjk FAILED                                  [ 25%]
tests/translate/storage/test_po.py::TestPOFile::test_wrap_parenthesis FAILED                              [ 50%]
tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrapping_cjk FAILED                              [ 75%]
tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrap_parenthesis FAILED                          [100%]

=================================================== FAILURES ====================================================
_________________________________________ TestPOFile.test_wrapping_cjk __________________________________________

self = <tests.translate.storage.test_po.TestPOFile object at 0x7fae5c8e71b0>

        def test_wrapping_cjk(self):
            posource = r"""msgid ""
    msgstr "Content-Type: text/plain; charset=utf-8\n"
    
    msgid "test"
    msgstr ""
    "効率的なバグの報告はPostGISの開発を助ける本質的な方法です。最も効率的なバグ報"
    "告は、PostGIS開発者がそれを再現できるようにすることで、それの引き金となったス"
    """
>           assert self.poreflow(posource) == posource
E           assert 'msgid ""\nms...れの引き金となったス"\n' == 'msgid ""\nms...れの引き金となったス"\n'
E             
E               msgid ""
E               msgstr "Content-Type: text/plain; charset=utf-8\n"
E               
E               msgid "test"
E               msgstr ""
E             + "効率的なバグの報告はPostGISの開発を助ける本質的な方法です。最も効率的なバグ報告は、PostGIS開発者がそれを再現できるようにすることで、それの引き金となったス"
E             - "効率的なバグの報告はPostGISの開発を助ける本質的な方法です。最も効率的なバグ報"
E             - "告は、PostGIS開発者がそれを再現できるようにすることで、それの引き金となったス"

tests/translate/storage/test_po.py:1052: AssertionError
_______________________________________ TestPOFile.test_wrap_parenthesis ________________________________________

self = <tests.translate.storage.test_po.TestPOFile object at 0x7fae5c8e7250>

        def test_wrap_parenthesis(self):
            posource = r"""msgid ""
    msgstr "Content-Type: text/plain; charset=utf-8\n"
    
    msgid "test"
    msgstr ""
    "Konvertiert [param what] in [param type] auf die bestmögliche Weise. Der "
    "[param type] verwendet die [enum Variant.Type]-Werte.\n"
    
    msgid "test2"
    msgstr ""
    "- Eine von [Object] abgeleitete Klasse, die in [ClassDB] existiert, z. B. "
    "[Node].\n"
    """
>           assert self.poreflow(posource) == posource
E           assert 'msgid ""\nms...[Node].\\n"\n' == 'msgid ""\nms...[Node].\\n"\n'
E             
E               msgid ""
E               msgstr "Content-Type: text/plain; charset=utf-8\n"
E               
E               msgid "test"
E               msgstr ""
E             - "Konvertiert [param what] in [param type] auf die bestmögliche Weise. Der "...
E             
E             ...Full output truncated (10 lines hidden), use '-vv' to show

tests/translate/storage/test_po.py:1078: AssertionError
________________________________________ TestPYPOFile.test_wrapping_cjk _________________________________________

self = <tests.translate.storage.test_pypo.TestPYPOFile object at 0x7fae5c88db30>

        def test_wrapping_cjk(self):
            posource = r"""msgid ""
    msgstr "Content-Type: text/plain; charset=utf-8\n"
    
    msgid "test"
    msgstr ""
    "効率的なバグの報告はPostGISの開発を助ける本質的な方法です。最も効率的なバグ報"
    "告は、PostGIS開発者がそれを再現できるようにすることで、それの引き金となったス"
    """
>           assert self.poreflow(posource) == posource
E           assert 'msgid ""\nms...れの引き金となったス"\n' == 'msgid ""\nms...れの引き金となったス"\n'
E             
E               msgid ""
E               msgstr "Content-Type: text/plain; charset=utf-8\n"
E               
E               msgid "test"
E               msgstr ""
E             + "効率的なバグの報告はPostGISの開発を助ける本質的な方法です。最も効率的なバグ報告は、PostGIS開発者がそれを再現できるようにすることで、それの引き金となったス"
E             - "効率的なバグの報告はPostGISの開発を助ける本質的な方法です。最も効率的なバグ報"
E             - "告は、PostGIS開発者がそれを再現できるようにすることで、それの引き金となったス"

tests/translate/storage/test_po.py:1052: AssertionError
______________________________________ TestPYPOFile.test_wrap_parenthesis _______________________________________

self = <tests.translate.storage.test_pypo.TestPYPOFile object at 0x7fae5c88dbd0>

        def test_wrap_parenthesis(self):
            posource = r"""msgid ""
    msgstr "Content-Type: text/plain; charset=utf-8\n"
    
    msgid "test"
    msgstr ""
    "Konvertiert [param what] in [param type] auf die bestmögliche Weise. Der "
    "[param type] verwendet die [enum Variant.Type]-Werte.\n"
    
    msgid "test2"
    msgstr ""
    "- Eine von [Object] abgeleitete Klasse, die in [ClassDB] existiert, z. B. "
    "[Node].\n"
    """
>           assert self.poreflow(posource) == posource
E           assert 'msgid ""\nms...[Node].\\n"\n' == 'msgid ""\nms...[Node].\\n"\n'
E             
E               msgid ""
E               msgstr "Content-Type: text/plain; charset=utf-8\n"
E               
E               msgid "test"
E               msgstr ""
E             - "Konvertiert [param what] in [param type] auf die bestmögliche Weise. Der "...
E             
E             ...Full output truncated (10 lines hidden), use '-vv' to show

tests/translate/storage/test_po.py:1078: AssertionError
============================================ short test summary info ============================================
FAILED tests/translate/storage/test_po.py::TestPOFile::test_wrapping_cjk - assert 'msgid ""\nms...れの引き金となったス"\n' == 'msgid ""\nms...れの引き金となったス"\n'
FAILED tests/translate/storage/test_po.py::TestPOFile::test_wrap_parenthesis - assert 'msgid ""\nms...[Node].\\n"\n' == 'msgid ""\nms...[Node].\\n"\n'
FAILED tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrapping_cjk - assert 'msgid ""\nms...れの引き金となったス"\n' == 'msgid ""\nms...れの引き金となったス"\n'
FAILED tests/translate/storage/test_pypo.py::TestPYPOFile::test_wrap_parenthesis - assert 'msgid ""\nms...[Node].\\n"\n' == 'msgid ""\nms...[Node].\\n"\n'
=============================================== 4 failed in 0.15s ===============================================
 
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -33,7 +33,8 @@ test = [
 ]
 types = [
   "setuptools>=61.2",
-  "types-lxml==2025.3.30"
+  "types-lxml==2025.3.30",
+  "wcwidth-stubs==0.2.14.0"
 ]
 
 [project]
@@ -61,8 +62,8 @@ classifiers = [
   "Topic :: Software Development :: Localization"
 ]
 dependencies = [
-  "cwcwidth>=0.1.10,<0.2",
-  "lxml>=5.2.0,<6.1"
+  "lxml>=5.2.0,<6.1",
+  "wcwidth>=0.2.14,<0.3"
 ]
 description = "Tools and API for translation and localization engineering."
 dynamic = [

--- a/tests/translate/storage/test_po.py
+++ b/tests/translate/storage/test_po.py
@@ -1198,7 +1198,6 @@ msgstr ""
         assert self.poreflow(gettext_0_23) == expected
         assert self.poreflow(gettext_0_20) == expected
 
-    @mark.xfail(reason="Incompatible wrapping with gettext, see #5661")
     def test_wrap_wide_stop(self):
         posource = r"""msgid ""
 msgstr "Content-Type: text/plain; charset=utf-8\n"

--- a/translate/storage/pypo.py
+++ b/translate/storage/pypo.py
@@ -105,7 +105,7 @@ def escapeforpo(line: str) -> str:
 def unicode_width(text: str) -> int:
     result = wcswidth(text)
     if result == -1:
-        raise ValueError("Unprintable character in a string")
+        raise ValueError("String contains control characters or non-displayable characters")
     return result
 
 

--- a/translate/storage/pypo.py
+++ b/translate/storage/pypo.py
@@ -31,6 +31,7 @@ import re
 import textwrap
 from functools import lru_cache
 from itertools import chain
+from warnings import warn
 
 from wcwidth import wcswidth
 
@@ -105,7 +106,10 @@ def escapeforpo(line: str) -> str:
 def unicode_width(text: str) -> int:
     result = wcswidth(text)
     if result == -1:
-        raise ValueError("String contains control characters or non-displayable characters")
+        warn(
+            f"String contains control characters or non-displayable characters: {text!r}"
+        )
+        return len(text)
     return result
 
 

--- a/translate/storage/pypo.py
+++ b/translate/storage/pypo.py
@@ -29,9 +29,10 @@ import copy
 import logging
 import re
 import textwrap
+from functools import lru_cache
 from itertools import chain
 
-from cwcwidth import wcswidth, wcwidth
+from wcwidth import wcswidth
 
 from translate.misc import quote
 from translate.misc.multistring import multistring
@@ -100,18 +101,30 @@ def escapeforpo(line: str) -> str:
     return po_escape_re.sub(escapehandler, line)
 
 
+@lru_cache(maxsize=128)
+def unicode_width(text: str) -> int:
+    result = wcswidth(text)
+    if result == -1:
+        raise ValueError("Unprintable character in a string")
+    return result
+
+
 def cjkslices(text: str, index: int) -> tuple[str, str]:
     """Return the two slices of a text cut to the index."""
-    text_width = wcswidth(text)
-    if text_width == len(text) or text_width <= index:
+    text_width = unicode_width(text)
+    # Do not wrap non-wide chars here
+    if text_width == len(text):
         return text, ""
-    length = 0
-    i = 0
-    for i in range(len(text)):
-        length += wcwidth(text[i])
-        if length > index:
-            break
-    return text[:i], text[i:]
+
+    # Guess split point
+    split_point = index * text_width // len(text)
+    while unicode_width(text[:split_point]) < index:
+        split_point += 1
+
+    while unicode_width(text[:split_point]) > index:
+        split_point -= 1
+
+    return text[:split_point], text[split_point:]
 
 
 class PoWrapper(textwrap.TextWrapper):
@@ -127,11 +140,13 @@ class PoWrapper(textwrap.TextWrapper):
         r"""
             (
             \[[^\]]{1,73}\]|                      # [] braces
+            \([^\)]{1,73}\)|                      # () braces
             \\"[^"]{1,73}\\"|                     # quoted string
             \s+|                                  # any whitespace
             [a-z0-9A-Z_#\[\].-]+/|                # nicely split long URLs
             \w*\\.\w*|                            # any escape should not be split
             [\w\!\'\&\.\,\?=<>%]+\s+|             # space should go with a word
+            [，。、]|                             # full width punctuation
             [^\s\w]*\w+[a-zA-Z]-(?=\w+[a-zA-Z])|  # hyphenated words
             (?<=[\w\!\"\'\&\.\,\?])-{2,}(?=\w)    # em-dash
             )
@@ -185,12 +200,21 @@ class PoWrapper(textwrap.TextWrapper):
             cur_len = 0
 
             while chunks:
-                l = wcswidth(chunks[-1])
+                l = unicode_width(chunks[-1])
 
                 # Can at least squeeze this chunk onto the current line.
                 if cur_len + l <= width:
                     cur_line.append(chunks.pop())
                     cur_len += l
+
+                # Figure out if we can split wide chars
+                elif l > len(chunks[-1]):
+                    slice1, slice2 = cjkslices(chunks[-1], width - cur_len)
+                    if slice2:
+                        cur_line.append(slice1)
+                        cur_len += unicode_width(slice1)
+                        chunks[-1] = slice2
+                    break
 
                 # Nope, this line is full.
                 else:
@@ -198,7 +222,7 @@ class PoWrapper(textwrap.TextWrapper):
 
             # The current line is full, and the next chunk is too big to
             # fit on *any* line (not just this one).
-            if not cur_line and chunks and wcswidth(chunks[-1]) > width:
+            if not cur_line and chunks and unicode_width(chunks[-1]) > width:
                 self._handle_long_word(chunks, cur_line, cur_len, width)
 
             if cur_line:

